Foundation/NSOrderedSet.swift

```diff
@@ -120,11 +121,26 @@ open class NSOrderedSet : NSObject, NSCopying, NSMutableCopying, NSSecureCoding,
         if type(of: self) === NSOrderedSet.self || type(of: self) === NSMutableOrderedSet.self {
             return _orderedStorage.map { _SwiftValue.fetch(nonOptional: $0) }
         } else {
-            return (0..<count).map { idx in
+            return _indices.map { idx in
                 return self[idx]
             }
         }
     }
+    
+    /// The range of indices that are valid for subscripting the ordered set.
+    internal var _indices: Range<Int> {
+        return 0..<count
+    }
+    
+    /// Checks that an index is valid: 0 ≤ `index` ≤ `count`.
+    internal func _validateIndex(_ index: Int, file: StaticString = #file, line: UInt = #line) {
+        precondition(index <= count && index >= 0, "\(self): Index out of bounds", file: file, line: line)
```

Comment:
Should this be `index < count`?

Reply:
Oh I see, it's used for insertion as well as removal. Those are two different scenarios I think.

Reply:
Correct. They are two different scenarios, which is why there are two different validation methods: `_validateIndex()` (which includes count as a valid index) and `_validateSubscript()`. This method (`_validateIndex()`) is only used for `insertObject()` and `setObject()`, both of which accept `count` as a valid index.

Reply:
It was the names that were confusing me; validate index sounds like it must be a valid current index, but really it's a valid current index or "future" index. 

Perhaps `_validatePossibleIndex` and `_validateCurrentIndex`; the latter could be used for more than just subscripting.